Session tracking

fractalwrench opened the pull request.

loopj commented on src/main/java/com/bugsnag/Configuration.java:
```diff
@@ -39,6 +50,8 @@
         if (ServletCallback.isAvailable()) {
             addCallback(new ServletCallback());
         }
+        sessionDelivery = new AsyncHttpDelivery();
```
Any reason you aren't initializing this at declaration time?

loopj commented on src/main/java/com/bugsnag/Configuration.java:
```diff
@@ -27,6 +37,7 @@
 
     Collection<Callback> callbacks = new ArrayList<Callback>();
     Serializer serializer = new Serializer();
+    private volatile boolean autoCaptureSessions;
```
`AtomicBoolean` might be better here.

loopj commented on src/main/java/com/bugsnag/Configuration.java:
```diff
@@ -39,6 +50,8 @@
             addCallback(new ServletCallback());
         }
+        sessionDelivery = new AsyncHttpDelivery();
+        ((HttpDelivery) sessionDelivery).setEndpoint("https://sessions.bugsnag.com");
```
We should probably have the default `setEndpoint` configured for sessions and notify in the same place. Currently we have `https://notify.bugsnag.com` set as the default error reporting endpoint in `SyncHttpDelivery.java` but session reporting defaults are configured here.

Also, should we have this URL in a a `static final` for easier visibility of default values all in one place?

loopj commented on src/main/java/com/bugsnag/ServletSessionTracker.java:
```diff
@@ -0,0 +1,36 @@
+ * {@link com.bugsnag.servlet.BugsnagServletRequestListener},
+ * intended for internal use
+ */
+public enum ServletSessionTracker {
```
Why an `enum` here?

fractalwrench replied:
`BugsnagServletRequestListener` can't take the `SessionTracker` as a constructor parameter, so the current solution was to track sessions statically via a singleton which holds a reference to the `SessionTracker`. According to Effective Java Item 3, an enum is the most foolproof way to create a singleton.

I've mentioned to @kattrali that I'm not entirely convinced that this is the right approach, particularly if a user has created multiple `Bugsnag` instances, as it would just track sessions in the most recently created object. I'm not particularly familiar with web.xml/EE so would be interested to hear anyone's thoughts on alternative approaches.

Pezzah commented on src/main/java/com/bugsnag/SessionTracker.java:
```diff
@@ -0,0 +1,90 @@
+
+                SessionPayload payload = new SessionPayload(requestValues, config);
+                Delivery delivery = config.sessionDelivery;
+                delivery.deliver(config.serializer, payload, config.getErrorApiHeaders());
```
This should use `config.getSessionApiHeaders()`

fractalwrench commented on src/main/java/com/bugsnag/SessionPayload.java:
```diff
@@ -0,0 +1,39 @@
+    private final Diagnostics diagnostics;
+
+    SessionPayload(Collection<SessionCount> sessionCounts, Configuration configuration) {
+        this.sessionCounts.addAll(sessionCounts);
```
I've removed this line so that `SessionPayload` just holds onto the reference, as the collection within the `SessionTracker` is already a defensive copy.

fractalwrench commented on src/main/java/com/bugsnag/Session.java:
```diff
@@ -0,0 +1,50 @@
+
+    Session(String id, Date startedAt) {
+        this.id = id;
+        this.startedAt = new Date(startedAt.getTime());
```
Unfortunately `Date` in Java is mutable. I've made a defensive copy here and in the getter to prevent anyone from altering the internals of the `Session` class.

fractalwrench commented on src/main/java/com/bugsnag/Bugsnag.java:
```diff
@@ -39,11 +59,45 @@ public Bugsnag(String apiKey, boolean sendUncaughtExceptions) {
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                sessionTracker.setShuttingDown(true);
```
That's correct. Looking back at this implementation, it might make sense to adhere to the following logic:

- Only collect a session in `startSession` if `shuttingDown` is false
- When `shuttingDown` is set to true, attempt to deliver any remaining sessions

What are your thoughts?

fractalwrench replied:
Updated to follow the logic laid out above.

fractalwrench commented on src/main/java/com/bugsnag/servlet/BugsnagServletRequestListener.java:
```diff
@@ -25,4 +30,12 @@ public void requestInitialized(ServletRequestEvent servletRequestEvent) {
     }
+
+    private void trackServletSession() {
+        for (Bugsnag bugsnag : Bugsnag.uncaughtExceptionClients()) {
```
If multiple bugsnag clients have been initialised, each of these will be informed that a HTTP request has been made, and record a session each if necessary. There should only ever be 1 session recorded for each client though.

Did you have specific concerns about performance here? This is definitely a hot path so it might be worth profiling, but at an initial glance I can't see too many expensive operations apart from possibly UUID generation/calendar usage.

I think this does mean that it can only track sessions when exceptions are being automatically caught.

fractalwrench replied:
This is definitely the behaviour we want - I've forwarded you an email with relevant design discussion

fractalwrench commented on src/main/java/com/bugsnag/SessionTracker.java:
```diff
@@ -0,0 +1,90 @@
+        boolean isNewBatchPeriod = isNewBatchPeriod(roundedStartDate);
+
+        if (isNewBatchPeriod) {
+            SessionCount prev = batchCount.getAndSet(new SessionCount(roundedStartDate));
```
Yes, I think so - nice spot!

fractalwrench commented:
I got around to addressing the review comments on this PR. I've also updated the `SessionTracker` to address a race condition that Martin spotted, so would be grateful if someone could have a thorough look at that, as it involves concurrency in a hot path.

fractalwrench commented on src/main/java/com/bugsnag/Configuration.java:
```diff
@@ -74,4 +88,28 @@ boolean inProject(String className) {
+
+    Map<String, String> getSessionApiHeaders() {
+        Map<String, String> map = new HashMap<String, String>();
+        map.put(HEADER_API_PAYLOAD_VERSION, "1.0");
```
The docs say 1.0 so I'm leaving this as is.

fractalwrench commented on src/main/java/com/bugsnag/Report.java:
```diff
@@ -114,6 +117,27 @@ public Map getMetaData() {
     }
 
+    @Expose
+    Map<String, Object> getSession() {
```
The `session` field isn't mutated here as far as I can see. This method builds a `Map` and is only ever invoked by the JSON serialiser.

fractalwrench commented on src/main/java/com/bugsnag/SessionTracker.java:
```diff
@@ -0,0 +1,93 @@
+class SessionTracker {
+
+    private final Configuration config;
+    private final ThreadLocal<Session> session = new ThreadLocal<Session>();
```
Generally speaking a request will be handled by a single thread, although there aren't any guarantees on _which_ thread that is. I've updated the JavaDoc for `startSession()` to make it obvious to anyone who is tracking sessions manually that we track on a per-thread basis.
